pykafka/producer.py

```diff
@@ -231,14 +231,32 @@ def _setup_owned_brokers(self):
                     self, partition.leader)
         return queued_messages
 
-    def stop(self):
-        """Mark the producer as stopped"""
-        self._running = False
+    def _stop_owned_brokers(self):
         self._wait_all()
         if self._owned_brokers is not None:
             for owned_broker in self._owned_brokers.values():
                 owned_broker.stop()
 
+    def stop(self, wait=True):
+        """Mark the producer as stopped, and wait until all messages to be sent
+
+        :param wait: whether we should wait until all messages to be sent or not
+        :type wait: bool
+        """
+        def get_queue_readers():
+            return [qr for qr in threading.enumerate() if qr.name.find('pykafka.OwnedBroker.queue_reader') == 0]
```

Comment:
This logic will break when `use_greenlets=True`. Two possible solutions occur to me. You could save references to the objects returned by `_cluster.handler.spawn` and pass them into a `self._cluster.handler.join_worker(t)` function, implementing both a gevent and a threaded version of that function. You could also write a `self._cluster.handler.get_workers_by_substring()` and call it here. The threaded version would use `threading.enumerate()`, and the greenlet version would use the gevent equivalent.

Also, I think that the identifying string for these worker threads should be deduplicated by putting it into a class-level variable.